ContainerEventProcessor internal segment could be evicted from memory

**Describe the bug**
During a experiment involving node reboots and container recoveries, we observed 2 containers on a single Segment Store continuously throwing `NullPointerException`s related to `StorageWriter`:
```
2021-10-01 14:33:01,358 82511409 [storage-io-36] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[10]: Iteration[5192372].Error.
java.lang.NullPointerException: null
2021-10-01 14:33:01,460 82511511 [storage-io-35] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[8]: Iteration[1743312].Error.
java.lang.NullPointerException: null
```
When looking at the origin of these persistent exceptions, we could see that the NPE in `StorageWriter` is related to the interactions with the `ContainerEventProcessor` (used for garbage collection purposes):
```
2021-09-30 17:33:45,846 6955897 [storage-io-130] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[8]: Iteration[1714522].Error.
java.lang.NullPointerException: null
	at io.pravega.segmentstore.server.containers.ContainerEventProcessorImpl$EventProcessorImpl.getOutstandingBytes(ContainerEventProcessorImpl.java:325)
	at io.pravega.segmentstore.server.containers.ContainerEventProcessorImpl$EventProcessorImpl.add(ContainerEventProcessorImpl.java:308)
	at io.pravega.segmentstore.server.containers.StorageEventProcessor.addTask(StorageEventProcessor.java:104)
	at io.pravega.segmentstore.server.containers.StorageEventProcessor.addTask(StorageEventProcessor.java:42)
	at io.pravega.segmentstore.storage.chunklayer.GarbageCollector.trackNewChunk(GarbageCollector.java:265)
	at io.pravega.segmentstore.storage.chunklayer.WriteOperation.addNewChunk(WriteOperation.java:280)
	at io.pravega.segmentstore.storage.chunklayer.WriteOperation.openChunkToWrite(WriteOperation.java:266)
	at io.pravega.segmentstore.storage.chunklayer.WriteOperation.lambda$writeData$11(WriteOperation.java:215)
	at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:848)
	at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:817)
	at io.pravega.common.concurrent.Futures.runOrFail(Futures.java:708)
	at io.pravega.common.concurrent.Futures$Loop.run(Futures.java:861)
	at java.base/java.util.concurrent.CompletableFuture$UniRun.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:189)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```
Checking why `ContainerEventProcessorImpl$EventProcessorImpl.getOutstandingBytes()` could be throwing NPE when attempting to add a new task. Apparently, the problem occurs because the `SegmentInfo` object is null after invoking: https://github.com/pravega/pravega/blob/1ac1a7acabadd8e3fa1bf979719dcb7d7f592d24/segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java#L324

In this sense, this is the first stacktrace where the `NullPointerException` error appears in `ContainerEventProcessor` for container 8:
```
2021-09-30 17:27:14,778 6564829 [ForkJoinPool.commonPool-worker-1] WARN  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[8-2]: Processing iteration failed, retrying.
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:189)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.NullPointerException: null
	at io.pravega.segmentstore.server.containers.StreamSegmentContainer.addOperation(StreamSegmentContainer.java:996)
	at io.pravega.segmentstore.server.containers.StreamSegmentContainer$DirectSegmentWrapper.truncate(StreamSegmentContainer.java:1087)
	at io.pravega.segmentstore.server.containers.ContainerEventProcessorImpl$EventProcessorImpl.truncateInternalSegment(ContainerEventProcessorImpl.java:496)
	at io.pravega.segmentstore.server.containers.ContainerEventProcessorImpl$EventProcessorImpl.lambda$processEvents$6(ContainerEventProcessorImpl.java:411)
	... 7 common frames omitted
```
As you can see, the NPE does not happen in `ContainerEventProcessor`, but in `StreamSegmentContainer` when the event processor tries to truncate the segment. Concretely, the problem is related to the `SegmentMetadata` being null for the segment that needs to be truncated:
```
    private <T extends Operation & SegmentOperation> CompletableFuture<Void> addOperation(T operation, Duration timeout) {
        SegmentMetadata sm = this.metadata.getStreamSegmentMetadata(operation.getStreamSegmentId()); <- sm is null here
        OperationPriority priority = calculatePriority(sm.getType(), operation);  <- NPE happens just here
        return this.durableLog.add(operation, priority, timeout);
    }
```
Before the NPE appears, I see this in the logs:
```
2021-09-30 17:27:06,788 6556839 [core-15] INFO  i.p.s.server.CacheManager - CacheManager Unregistered ReadIndex[8-2] (_system/containers/event_processor_GC.queue.8_8).
```
This may indicate that the internal segment for `ContainerEventProcessor` has been evicted. However, `ContainerEventProcessor` is intended to work with internal, pinned, system-critical segments which should never be evicted from memory: https://github.com/pravega/pravega/blob/1ac1a7acabadd8e3fa1bf979719dcb7d7f592d24/segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java#L129

**To Reproduce**
Problem observed while running a longevity workload and doing a node reboot experiment.

**Screenshots**
n/a

**Additional information**
n/a

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java
@@ -30,7 +30,6 @@ import io.pravega.common.io.serialization.VersionedSerializer;
 import io.pravega.common.util.BufferView;
 import io.pravega.common.util.ByteArraySegment;
 import io.pravega.segmentstore.contracts.SegmentType;
-import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
 import io.pravega.segmentstore.server.ContainerEventProcessor;
 import io.pravega.segmentstore.server.DirectSegmentAccess;
 import io.pravega.segmentstore.server.SegmentContainer;
@@ -123,15 +122,13 @@ class ContainerEventProcessorImpl implements ContainerEventProcessor {
      * @return A future that, when completed, contains reference to the Segment to be used by a given
      * {@link ContainerEventProcessor.EventProcessor} based on its name.
      */
-    private static Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
+    @VisibleForTesting
+    static Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
                                                                                                        MetadataStore metadataStore,
                                                                                                        Duration timeout) {
-        return s -> Futures.exceptionallyComposeExpecting(
-                container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout),
-                e -> e instanceof StreamSegmentNotExistsException,
-                () -> metadataStore.registerPinnedSegment(getEventProcessorSegmentName(container.getId(), s),
-                        SYSTEM_CRITICAL_SEGMENT, null, timeout) // Segment should be pinned.
-                        .thenCompose(l -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout)));
+        return s -> metadataStore.registerPinnedSegment(getEventProcessorSegmentName(container.getId(), s),
+                SYSTEM_CRITICAL_SEGMENT, null, timeout) // Segment should be pinned.
+                .thenCompose(l -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout));
     }
 
     //endregion

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/containers/StreamSegmentContainerTests.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/containers/StreamSegmentContainerTests.java
@@ -160,6 +160,7 @@ import org.junit.rules.Timeout;
 
 import static io.pravega.common.concurrent.ExecutorServiceHelpers.newScheduledThreadPool;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
 
 /**
  * Tests for StreamSegmentContainer class.
@@ -2491,6 +2492,25 @@ public class StreamSegmentContainerTests extends ThreadPooledTestSuite {
         ContainerEventProcessor containerEventProcessor = new ContainerEventProcessorImpl(container, container.metadataStore,
                 TIMEOUT_EVENT_PROCESSOR_ITERATION, TIMEOUT_EVENT_PROCESSOR_ITERATION, this.executorService());
         ContainerEventProcessorTests.testEventRejectionOnMaxOutstanding(containerEventProcessor);
+    }
+
+    /**
+     * Tests that call to getOrCreateInternalSegment is idempotent and always provides pinned segments.
+     */
+    @Test(timeout = 30000)
+    public void testPinnedSegmentReload() {
+        @Cleanup
+        TestContext context = createContext();
+        val container = (StreamSegmentContainer) context.container;
+        container.startAsync().awaitRunning();
+        Function<String, CompletableFuture<DirectSegmentAccess>> segmentSupplier =
+                ContainerEventProcessorImpl.getOrCreateInternalSegment(container, container.metadataStore, TIMEOUT_EVENT_PROCESSOR_ITERATION);
+        long segmentId = segmentSupplier.apply("dummySegment").join().getSegmentId();
+        for (int i = 0; i < 10; i++) {
+            DirectSegmentAccess segment = segmentSupplier.apply("dummySegment").join();
+            assertTrue(segment.getInfo().isPinned());
+            assertEquals(segmentId, segment.getSegmentId());
+        }
     }
 
     /**

